### ISSUE TITLE
how can a client detect if IDs should be old style or new FDSN style

### ISSUE THREAD

**Issue by crotwell** [EarthScope/ringserver#61]
new issue (from #60)

Conern that v4 ringserver cannot be drop in replacement for prev version due to auto-upgrade of packet ids to fdsn sourceid style for datalink. While I like new style source ids and believe they should be used, the auto-migration of ids (I think) means that while data input to new ringserver is ok regardless of how the client creates ids, the datalink output side breaks backwards compatibility. 

Should old datalink clients be able to connect to get data without changing their MATCH pattern?

For datalink with old style ids it doesn't find any streams. So this works:
```
Send: MATCH FDSN:CO_JSC_00_H_H_./MSEED 
start datalink at : 2024-12-15T14:20:12.776Z 
Send: POSITION AFTER 1734272412776000 undefined 
Send: STREAM
```
but this fails to match any streams:
```
Send: MATCH CO_JSC_00_HH./MSEED 
```

The release notes sound like old style match patterns for datalink should still work. Am I misunderstanding?

> All legacy stream IDs for miniSEED of the form: NN_SSSSS_LL_CCC/MSEED will be converted to FDSN Source ID form: FDSN:NN_SSSSS_LL_C_C_C/MSEED.

**Comment by chad-earthscope** [comment]
You are not misunderstanding.  Version 4 is a new major version that I've tried to make as drop-in-replacement as possible with DataLink stream IDs being an exception.

In the general, ringserver stream IDs are arbitrary with a convention for miniSEED records.  Adopting and providing a transition for new, standard-based convention from the old convention makes a lot of sense.  Since SeedLink does not directly select stream IDs, the SeedLink selection syntax can be transparently mapped to the new miniSEED convention, which is much easier when the ID convention is consistent (thus the translation of old to new on insertion).  DataLink, OTOH, directly matches stream IDs and is not trivially adapted to the new convention.  The use case of ringserver as miniSEED-for-SeedLink received the most consideration regarding backwards compatibility while moving to a new convention.  I'm sure you agree this makes sense.

So why didn't DataLink get a transparent path to the new convention?  This is mostly due to my estimation that the vast majority of use cases do not include using DataLink as a consuming client, combined with the complexity of building that new path.  You are a noted exception to using DataLink as a consuming client and you've promoted that path with supporting software (which I appreciate!).  I'm hopeful that most users and use cases of ringserver will not be effected in this major version change.

I'd like to provide a transition path for DataLink clients selecting streams using the old, ringserver-specific, convention if I could figure out how.  I've considered these options:

1) translate the submitted MATCH/REJECT regexes to the new convention when they match the old (which is what is happening for SeedLink).  This would be really difficult to get right.  I'm doubt it's even possible as I expect there are many, many ambiguities in trying to convert arbitrarily complex regexes to a new form.

2) restructure the ring buffer to allow streams to be identified with multiple IDs, likely by changing fundamental stream IDs to UUIDs/hash values and some lookup to match 1-N text IDs.  This is non-trivial complexity.

Fundamentally it comes down to "is this worth the work and complexity".  If I could find an easy way, or a lot of spare time, I'd make this work.  Other suggestions welcome.

P.S. Note on the option to restructure the ring buffer (2) for future consideration: replace the internal stream index (the `streamidx` file) with an sqlite3 DB.  Including an extra table that translates stream UUID/hash to text stream IDs would be relatively straightforward.  This is a non-trivial change.

**Comment by chad-earthscope** [comment]
There are other considerations for DataLink clients.  If the selection was translated would the packet be returned as the translated ID or one of it's alternates?

**Comment by crotwell** [comment]
Humm, yes rewritting a regex would be challenging. I'm inclined to agree with the breaking-compatibility decision.

Change of question, how can a client detect if IDs should be old style or new FDSN style?  I would like to create example clients that will work today, but also will not break on the day rtserver.iris upgrades to v4. Is there something in the initial ID handshake that would trigger this for a client that wished to interoperate with both old and v4 ringserver,ie it knows station and channels and wishes to pass in correct MATCH regex for them? Is keying off of something in ID best, like:
```
DataLink v1.1 (RingServer/4.0.0) :: DLPROTO:1.1 PACKETSIZE:512 WRITE
```
vs
```
DataLink 2018.078 :: DLPROTO:1.0 PACKETSIZE:512
```

Guessing the `DLPROTO:1.1` is best? Or would it be worth having an additional "capability" in the ID line to indicate ID style? Not sure there is an easy way for a client to create a regex that works in both cases. Regardless, documenting an upgrade path for clients would be helpful.

BTW, what is difference between `DLPROTO:1.1` and `DLPROTO:1.0`? I only see docs for the 2008 version here:
https://earthscope.github.io/libdali/datalink-protocol.html
Would be nice if the protocol docs could be updated to show 1.0 vs 1.1 differences.

**Comment by chad-earthscope** [comment]
> Guessing the DLPROTO:1.1 is best

Yes, that's probably the best option.  It's probably best to do `DLPROTO:1.0` versus any later version.  Also keep in mind that the stream IDs are arbitrary and an operator can insert whatever they want.  ringserver v4 will stream anything it detects as miniSEED via SeedLink regardless of stream ID.

**Comment by chad-earthscope** [comment]
>  ringserver v4 will stream anything it detects as miniSEED via SeedLink regardless of stream ID.

Hmm, well this is probably true but with the important caveat that SeedLink v4 station ID is generated from the ringserver stream ID and is expected to be the FDSN Source ID.  If it's not a Source ID it'll use the whole stream ID for the SeedLink v4 "station ID" field.

**Comment by crotwell** [comment]
Can the `DLPROTO` value be discovered via http for non-datalink uses? Normally, I would say datalink things should be access via a datalink websocket, but the format of the stream ids is important for example when getting stream latency for a known NSLC channel via http. You have to know how to form the `match` query param based on the server and not clear how to get this information just via http.

I guess one could key off of the returned value from /id, like:
```
ringserver/4.0.1
Organization: South Carolina Seismic Network
Server start: 2025-01-09T18:03:39Z
```
and assume that version >4 means `FDSN:` style ids. Is the best way?

Or key off of `Ring version: 2` in `/status`?

Would it be worth adding Seedlink and Datalink "capabilites" in the `/status` output?

**Comment by chad-earthscope** [comment]
I'm thinking a list of supported protocols, and a boolean `fdsn_source_identifiers` should be added to `/status/json` (and the formatted output.

**Comment by crotwell** [comment]
Trying latest develop branch, so should be basically 4.1rc, but /status and /status/json both give a 403 from a non-local ip. Output from ringserver is:

> HTTP STATUS request from un-trusted client

If /status is how a client determines what style identifiers to use, then how does that work if it requires trust? Seems like that info needs to be as public as /streams and the data?

**Comment by chad-earthscope** [comment]
Hmm, right, that endpoint is only accessible by trusted clients.  So this would need to go into another endpoint like `/id` or a new one like `/capabilities`

**Comment by crotwell** [comment]
Either seems fine, or both? 

I guess `/id` has the advantage that it already exists on prior versions, so client doesn't have to guess if a 404 on `/capabilities` means older ringserver or just a proxy misconfiguration. On other hand I kind of like cleanliness of `/capabilities`.

Related, `Server start:` in `/id` seems more like a `/status` thing. Is there a reason that is public? Just debugging?

**Comment by chad-earthscope** [comment]
The `develop` branch now has them at the /id level, so they will show up on /status response as well.

> Related, Server start: in /id seems more like a /status thing. Is there a reason that is public? Just debugging?

Historical quirk of being included in SeedLink ID INFO response 20+ years ago.  Not worth changing IMHO.

**Comment by crotwell** [comment]
My old ringserver /id shows:
```
ringserver/2020.075
Organization: South Carolina Seismic Network
```
but the new one shows 
```
ringserver/4.1.0-RC
Organization: Test Ring Server 4
Server start: 2025-05-12T14:49:04Z
```

Has `Server start` really been there for 20 years?

**Comment by chad-earthscope** [comment]
Yes, the server start has been included in the SeedLink ID INFO response for 20+ years.  There was only one server implementation back then and it included it.

ringserver now prints it via HTTP for consistency of the "ID" elements reported, i.e. it's in the SeedLink response along with the server protocols now.

This ticket topic is now complete.
